encoder/flatfile.go

```diff
@@ -0,0 +1,225 @@
+}
+
+func (c *FlatFileItemWriter) Close() error {
+	return c.writer.Close()
+}
+func (c *FlatFileItemWriter) Write(item common_datalayer.Item) error {
+	buf := new(bytes.Buffer)
+	row := item.NativeItem().(map[string]any)
+	line := make([]string, 0)
+	var preppedValue string
+	var fieldValue interface{}
+	fieldsWithData := 0
+	for i, field := range c.fields {
+
+		if _, ok := row[field.Name]; ok {
+
+			fieldConfig := c.fields[i]
+			fieldValue = row[field.Name]
+			fieldSize := fieldConfig.Length
+			if fieldValue == nil {
+				//	Need to add spaces according to field length config
+				preppedValue = appendSpaces(preppedValue, fieldSize)
+			} else {
+				//	cast to string, then cut or append spaces to value according to config
+				var valueLength int
+				switch fieldValue.(type) {
+				case float64:
+					fieldValue = fmt.Sprintf("%f", fieldValue)
+					fieldValue = strings.Replace(fieldValue.(string), ".", "", -1)
+					valueLength = len(fieldValue.(string))
+				default:
+					fieldValue = fmt.Sprintf("%v", fieldValue)
+				}
+				valueLength = len(fieldValue.(string))
+				// do this in mapper?
+				if valueLength < fieldSize {
+					diff := fieldSize - valueLength
+					if fieldConfig.Type == "int" {
+						preppedValue = prependZeros(fieldValue.(string), diff)
+					} else {
+						preppedValue = appendSpaces(fieldValue.(string), diff)
+					}
+				} else if valueLength > fieldSize {
+					preppedValue = fieldValue.(string)[:fieldSize]
+				} else {
+					preppedValue = fieldValue.(string)
+				}
+				fieldsWithData += 1
+			}
+		}
+		line = append(line, preppedValue)
+	}
+	if fieldsWithData != 0 {
+		buf.WriteString(fmt.Sprintf("%s\n", strings.Join(line, "")))
+	}
+	_, err := c.writer.Write(buf.Bytes())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type FlatFileItemIterator struct {
+	reader    io.ReadCloser
+	scanner   *bufio.Scanner
+	fields    []FlatFileField
+	indexFrom int
+}
+
+func NewFlatFileItemIterator(sourceConfig map[string]any, data io.ReadCloser) (*FlatFileItemIterator, error) {
+	scanner := bufio.NewScanner(data)
+	var indexFrom = 0
+	if sourceConfig["indexFrom"] != nil {
+		indexFrom = sourceConfig["indexFrom"].(int)
+	}
+	reader := &FlatFileItemIterator{reader: data, scanner: scanner, indexFrom: indexFrom}
+	itemReader, err2 := NewFlatFileReadConfig(sourceConfig["fields"].([]map[string]interface{}), reader)
+	if err2 != nil {
+		return itemReader, err2
+	}
+
+	return reader, nil
+}
+
+func (c *FlatFileItemIterator) Close() error {
+	return c.reader.Close()
+}
+
+func (c *FlatFileItemIterator) Read() (common_datalayer.Item, error) {
+	var entityProps = make(map[string]interface{})
+	for c.scanner.Scan() {
+		line := c.scanner.Text()
+
+		var step = 0
+		if c.indexFrom != 0 {
+			step = c.indexFrom
+		}
+		for _, field := range c.fields {
+			if field.Ignore {
+				step += field.Length
+				continue
+			} else {
+				entityProps[field.Name] = strings.TrimSpace(line[step : step+field.Length])
```

Comment:
Would be tempted here to create an item and then use the functions of Item to set the values. Allows, if needed, for impl of Item to change and not affect this part.

Reply:
leaving for now